core/network/impl/state_protocol_observer_impl.cpp

```diff
@@ -0,0 +1,159 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "network/impl/state_protocol_observer_impl.hpp"
+#include <boost/bind/storage.hpp>
+#include <libp2p/outcome/outcome.hpp>
+
+#include "blockchain/block_header_repository.hpp"
+#include "common/buffer.hpp"
+#include "network/types/state_response.hpp"
+#include "storage/predefined_keys.hpp"
+#include "storage/trie/trie_storage.hpp"
+#include "storage/trie/types.hpp"
+
+constexpr unsigned MAX_RESPONSE_BYTES = 2 * 1024 * 1024;
+
+OUTCOME_CPP_DEFINE_CATEGORY(kagome::network,
+                            StateProtocolObserverImpl::Error,
+                            e) {
+  using E = kagome::network::StateProtocolObserverImpl::Error;
+  switch (e) {
+    case E::INVALID_CHILD_ROOTHASH:
+      return "Expected child root hash prefix.";
+    case E::NOTFOUND_CHILD_ROOTHASH:
+      return "Child storage root hash not found.";
+  }
+  return "unknown error";
+}
+
+namespace kagome::network {
+
+  StateProtocolObserverImpl::StateProtocolObserverImpl(
+      std::shared_ptr<blockchain::BlockHeaderRepository> blocks_headers,
+      std::shared_ptr<storage::trie::TrieStorage> storage)
+      : blocks_headers_(std::move(blocks_headers)),
+        storage_{std::move(storage)},
+        log_(log::createLogger("StateProtocolObserver", "network")) {
+    BOOST_ASSERT(blocks_headers_);
+    BOOST_ASSERT(storage_);
+  }
+
+  outcome::result<std::pair<KeyValueStateEntry, size_t>>
+  StateProtocolObserverImpl::getEntry(const storage::trie::RootHash &hash,
+                                      const common::Buffer &key,
+                                      size_t limit) const {
+    OUTCOME_TRY(batch, storage_->getEphemeralBatchAt(hash));
+
+    auto cursor = batch->trieCursor();
+
+    KeyValueStateEntry entry;
+    entry.state_root = hash;
+
+    auto res = key.empty() ? cursor->next() : cursor->seekUpperBound(key);
+    size_t size = 0;
+    if (res.has_value()) {
+      while (cursor->key().has_value() && size < limit) {
+        if (auto value_res = batch->tryGet(cursor->key().value());
+            value_res.has_value()) {
+          const auto &value = value_res.value();
+          entry.entries.emplace_back(
+              StateEntry{cursor->key().value(), value.value()});
```

Comment:
Value is optional and might be nullopt. Please, cover this case or add assertion.

Reply:
I think nullopt value in state query is some kind of error in state trie, it is inserted in logic artificially, it is not even an empty value. I just make fixes to skip such values, but I think, that it's a signal that node is in failed state.